### ISSUE TITLE
Support pre-split microbatches in distributed pipeline parallelism

### ISSUE THREAD

**Issue by acisseJZhong** [pytorch/torchtitan#2370]
Today in Torchtitan PP takes in full batch [data](https://github.com/pytorch/torchtitan/blob/e38d7abe0fa09c2ec7ccac1729f77cf72faddfc5/torchtitan/train.py#L507), automatically [split the batches](https://github.com/pytorch/pytorch/blob/03406903616077227734f772d682fc6027513ecf/torch/distributed/pipelining/schedules.py#L638) into microbatches, and does forward backward for you; in my use case, I would like to
1. for PP schedule to take in pre-split microbatches(unpadded, so PP would need to handle the padding)
2. have control to do forward only

Had a very [draft PR](https://github.com/acisseJZhong/verl/pull/1/changes#diff-f658afe18d14b480f4067f7544fbdb0ef6962a20ef3b5f5d0c709ae31e91809dR700) with claude to hack PP using only [`_step_microbatches`](https://github.com/pytorch/pytorch/blob/03406903616077227734f772d682fc6027513ecf/torch/distributed/pipelining/schedules.py#L647)`; but it's not runnable/correct yet.

**Comment by tianyu-l** [comment]
cc @H-Huang

**Comment by H-Huang** [comment]
Hey @acisseJZhong, trying to better understand the gaps. For 2) The schedules have [`eval()`](https://github.com/pytorch/pytorch/blob/89f3102643c5dfca19af7adeb6e0332a7c9b68fb/torch/distributed/pipelining/schedules.py#L365) to run forward only, lmk if that is not enough.

For 1), what is stopping you from pre-splitting the microbatches ahead of time and passing it into the schedule step? The logic of the schedule microbatch splitting goes through `split_args_kwargs_into_chunks`, so here is an example demonstrate what I mean:

```python
import torch
from torch.distributed.pipelining.microbatch import split_args_kwargs_into_chunks

args = (torch.arange(12).reshape(4, 3),)
chunks = 2

def split_and_print(args, chunks=2):
    args_split, _ = split_args_kwargs_into_chunks(args, None, chunks)
    print(f"Number of chunks: {len(args_split)}")
    for i, a in enumerate(args_split):
        print(f"\n--- Chunk {i} ---")
        for j, tensor in enumerate(a):
            print(f"  args[{j}] shape: {tensor.shape}")
    return args_split


# Original: pass full tensors, let the function split them
print("=== Using split_args_kwargs_into_chunks ===")
args_split_orig = split_and_print(args, chunks=chunks)

# Pre-split: chunk the tensors manually, still pass into split_args_kwargs_into_chunks
print("\n=== Pre-split manually ===")
pre_split_args = tuple(t.chunk(chunks) for t in args)
pre_split_args_split_orig = split_and_print(args, chunks=chunks)

print(f"{args_split_orig=}")
print(f"{pre_split_args_split_orig=}")
```

Output:
```
=== Using split_args_kwargs_into_chunks ===
Number of chunks: 2

--- Chunk 0 ---
  args[0] shape: torch.Size([2, 3])

--- Chunk 1 ---
  args[0] shape: torch.Size([2, 3])

=== Pre-split manually ===
Number of chunks: 2

--- Chunk 0 ---
  args[0] shape: torch.Size([2, 3])

--- Chunk 1 ---
  args[0] shape: torch.Size([2, 3])

args_split_orig=[(tensor([[0, 1, 2],
        [3, 4, 5]]),), (tensor([[ 6,  7,  8],
        [ 9, 10, 11]]),)]
pre_split_args_split_orig=[(tensor([[0, 1, 2],
        [3, 4, 5]]),), (tensor([[ 6,  7,  8],
        [ 9, 10, 11]]),)]
```

In both cases, you end up with the same chunks. I am just adding the `pre_split_args = tuple(t.chunk(chunks) for t in args)` to pre-split it manually and save it into a tuple. Then this gets passed into `split_args_kwargs_into_chunks` to split up this tuple.

**Comment by acisseJZhong** [comment]
got it, let me try passing a tuple of pre-split microbatches into `step/eval`. In this case, I'd imagine I need to pad the microbatches to max `seq_len` ahead of time? 

Also curious how we should pass in [`extra_kwargs` and `extra_inputs`](https://github.com/pytorch/torchtitan/blob/a103f4c5e0b71667dce111437ad4a90308640746/torchtitan/train.py#L508), I have attention_masks in extra kwargs, and positions in extra inputs. From[ this function ](https://github.com/pytorch/pytorch/blob/03406903616077227734f772d682fc6027513ecf/torch/distributed/pipelining/schedules.py#L466)it doesn't support passing in pre-split extra kwargs? 

specifically, this is [how I tried using eval](https://github.com/acisseJZhong/verl/blob/dbbad4cdd4dcbae7925cdef09333a678c5864091/verl/workers/engine/torchtitan/transformer_impl.py#L884) and [preparing a tuple of mbs ](https://github.com/acisseJZhong/verl/blob/dbbad4cdd4dcbae7925cdef09333a678c5864091/verl/workers/engine/torchtitan/transformer_impl.py#L713)

**Comment by H-Huang** [comment]
Yeah, good points. For padding, that is not supported yet so let me think about how we can enable that. UI-wise, i'm not sure whether we would want to offer this by default or just raise an error if that dim is not evenly split by microbatches.

For kwargs, I am not sure. `_split_inputs` which calls into `split_args_kwargs_into_chunks` should still also split the kwargs, but there may be some gaps. I'll also test that out.

**Comment by acisseJZhong** [comment]
> For padding, that is not supported yet so let me think about how we can enable that. UI-wise, i'm not sure whether we would want to offer this by default or just raise an error if that dim is not evenly split by microbatches.

yeah I think there are two aspects: 
1) for uneven microbatch size split, this is currently not supported right? We can only pass in microbatches with the same batch size(which is the case for me, as I pre-split microbatches and pack them into bz1, so each microbatch will have bz=1). 

2) for padding max `seq_len`, each microbatch has shape `[bz, seq_len]` with different `seq_len` and I need to pad to max_seq before passing into `eval`. As a consequence, I would also need to handle padding for other inputs like `positions`,  `attention_mask`, and `targets` and pass them as a list/tuple of dictionarys(kwargs was just a dictionary).

**Comment by wwwjn** [comment]
Related RP: https://github.com/pytorch/torchtitan/pull/3856 this would help integration with VeRL too, right? cc @acisseJZhong

**Comment by tianyu-l** [comment]
Yes, but I think Megatron in general supports variable length microbatches, which in the first version we won't. @sanketpurandare

**Comment by acisseJZhong** [comment]
yes verl PP should be unblocked. I am not prioritizing supporting PP in Verl yet since no oss users is asking for it anyways.

**Comment by sanketpurandare** [comment]
> Yes, but I think Megatron in general supports variable length microbatches, which in the first version we won't. [@sanketpurandare](https://github.com/sanketpurandare)

Yes, the first version only supports pre-split microbatches requiring them to be of same size. Supporting variable length microbatches is a design question. Do we want to support a max microbatch size and use padding during compute or do we want to only do padding and unpadding at the stage boundaries during send/recv. I am only willing to invest in this if there is a real use-case and requirement.
